spanner/spanner_snippets/spanner/spanner_create_backup_with_encryption_key.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_create_backup_with_encryption_key]
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	pbt "github.com/golang/protobuf/ptypes/timestamp"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+func createBackupWithCustomerManagedEncryptionKey(w io.Writer, db, backupID, kmsKeyName string) error {
+	// kmsKeyName = `projects/<project>/locations/<location>/keyRings/<key_ring>/cryptoKeys/<kms_key_name>`
+	matches := regexp.MustCompile("^(.+)/databases/(.+)$").FindStringSubmatch(db)
+	if matches == nil || len(matches) != 3 {
+		return fmt.Errorf("createBackupWithCustomerManagedEncryptionKey: invalid database id %q", db)
+	}
+	instanceName := matches[1]
+
+	ctx := context.Background()
+	adminClient, err := database.NewDatabaseAdminClient(ctx)
+	if err != nil {
+		return fmt.Errorf("createBackupWithCustomerManagedEncryptionKey.NewDatabaseAdminClient: %v", err)
+	}
+	defer adminClient.Close()
+
+	expireTime := time.Now().AddDate(0, 0, 14)
+	// Create a backup for a database using a Customer Managed Encryption Key
+	req := adminpb.CreateBackupRequest{
+		Parent:   instanceName,
+		BackupId: backupID,
+		Backup: &adminpb.Backup{
+			Database:   db,
+			ExpireTime: &pbt.Timestamp{Seconds: expireTime.Unix(), Nanos: int32(expireTime.Nanosecond())},
+		},
+		EncryptionConfig: &adminpb.CreateBackupEncryptionConfig{
+			KmsKeyName:     kmsKeyName,
+			EncryptionType: adminpb.CreateBackupEncryptionConfig_CUSTOMER_MANAGED_ENCRYPTION,
+		},
+	}
+	op, err := adminClient.CreateBackup(ctx, &req)
+	if err != nil {
+		return fmt.Errorf("createBackupWithCustomerManagedEncryptionKey.CreateBackup: %v", err)
+	}
+	// Wait for backup operation to complete.
+	backup, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("createBackupWithCustomerManagedEncryptionKey.Wait: %v", err)
+	}
+
+	// Get the name, create time, backup size and encryption key from the backup.
+	backupCreateTime := time.Unix(backup.CreateTime.Seconds, int64(backup.CreateTime.Nanos))
+	fmt.Fprintf(w,
+		"Backup %s of size %d bytes was created at %s using encryption key %s\n",
+		backup.Name,
+		backup.SizeBytes,
+		backupCreateTime.Format(time.RFC3339),
+		kmsKeyName)
```

Comment:
For the key name, you should be able to get it from the backup object to make sure that what's used for encryption is the same as what you specified as the key.

Reply:
I can't see a field on the `Backup` object that contains a field for that. It does contain an `EncryptionInfo` [struct](https://github.com/googleapis/go-genproto/blob/93e7006c17a68b79787d42654b179335e7337138/googleapis/spanner/admin/database/v1/backup.pb.go#L224), which again [contains](https://github.com/googleapis/go-genproto/blob/93e7006c17a68b79787d42654b179335e7337138/googleapis/spanner/admin/database/v1/common.pb.go#L238) a `KmsKeyVersion` field. We could use that, but at least for `Database` objects it [might take a little while](https://github.com/googleapis/googleapis/blob/5da78db589d423db6bcd6a35637c923fe1a7665c/google/spanner/admin/database/v1/spanner_database_admin.proto#L399) before it is is populated, and I'm not sure whether that is also true for backups.

We could also use the key name from the `CreateBackupRequest` object, but that is not really different from the `kmsKeyName` variable.

Or am I missing something else here?

Reply:
Okay, this confusion is my fault. My previous comment on this PR was incorrect. I hadn't realized that `Backup` did not have `encryption_config`.

For `Database`, we should use `encryption_config` as `encryption_info` is lazily instantiated.

For `Backup`, there is no `encryption_config`. The `encryption_info` field on `Backup` does not mention being lazily instantiated and does appear to be set every time. However, using `encryption_info` means the string will include the key version which is inconsistent with the other sample messages.

The two options are:
1) using the `kmsKeyName` i.e. no change
2) using the `EncryptionInfo.kmsKeyVersion`
I need to follow up to see if it's possible to use `EncryptionInfo.kmsKeyVersion` with the database samples for consistency. I'll get back to you in the next day or so :+1:

Reply:
Thanks for the clarification :+1: . I just changed it to using `EncryptionInfo.KmsKeyVersion` after having a quick look at the NodeJS sample. It seems that it is always populated (at least the tests succeed there). Changing it back to using the `kmsKeyName` is simple, so I'll leave it like this until we know which one we prefer.